### PR TITLE
Add JSON Schemas for API objects

### PR DESCRIPTION
**Pull request by mjrosengrant**
- Adds optional dependency marshmallow

### CODE REVIEW

File: testing/schema_tests.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+        self.assertEqual(menu_btn.outline.color.selected.hex, test_menu_btn.outline.color.selected.hex)
+        self.assertEqual(menu_btn.outline.color.unusable.hex, test_menu_btn.outline.color.unusable.hex)
+        self.assertEqual(menu_btn.outline.size.idle, test_menu_btn.outline.size.idle)
+        self.assertEqual(menu_btn.outline.size.highlighted, test_menu_btn.outline.size.highlighted)
+        self.assertEqual(menu_btn.outline.size.selected, test_menu_btn.outline.size.selected)
+        self.assertEqual(menu_btn.outline.size.unusable, test_menu_btn.outline.size.unusable)
+
+    def test_dump_menu(self):
+        """Ensure that dumping menu from serializers returns same input json."""
+        # self.maxDiff = None
```
Comment (fiskenslakt) [inline testing/schema_tests.py:77]: Are you leaving this in?

**Review by fiskenslakt — APPROVED** [review]
Looks damn good to me

File: nanome/api/schemas/ui_schemas.py [hunk 2]
```diff
@@ -0,0 +1,467 @@
...
+                setattr(new_obj, key, data[key])
+            except AttributeError:
+                raise AttributeError('Could not set attribute {}'.format(key))
+        return new_obj
+
+
+class ImageSchema(Schema):
+    type_name = fields.String(required=True)
+    color = ColorField()
+    file_path = fields.String()
```
Comment (ajm13) [inline nanome/api/schemas/ui_schemas.py:246]: what about image data?

File: nanome/api/schemas/ui_schemas.py [hunk 3]
```diff
@@ -0,0 +1,467 @@
...
+    text = fields.String(attribute='text_value')
+    text_vertical_align = EnumField(enum=enums.VertAlignOptions)
+    text_horizontal_align = EnumField(enum=enums.HorizAlignOptions)
+    text_auto_size = fields.Bool()
+    text_max_size = FloatRoundedField()
+    text_min_size = FloatRoundedField()
+    text_size = FloatRoundedField()
+    text_color = ColorField()
+    text_bold = fields.Bool()
+    text_italics = fields.Bool()
```
Comment (ajm13) [inline nanome/api/schemas/ui_schemas.py:289]:
```suggestion
    text_italic = fields.Bool()
```

File: nanome/api/schemas/ui_schemas.py [hunk 4]
```diff
@@ -0,0 +1,467 @@
...
+                raise AttributeError('Could not set attribute {}'.format(key))
+        return new_obj
+
+
+class SliderSchema(Schema):
+    type_name = fields.String(required=True)
+    current_value = FloatRoundedField()
+    min_value = FloatRoundedField()
+    max_value = FloatRoundedField()
+    type_name = fields.String()
```
Comment (ajm13) [inline nanome/api/schemas/ui_schemas.py:338]:
```suggestion
```
duplicate

File: nanome/api/schemas/ui_schemas.py [hunk 5]
```diff
@@ -0,0 +1,467 @@
...
+    selected = fields.Bool()
+
+
+class DropdownSchema(Schema):
+    type_name = fields.String(required=True)
+    permanent_title = fields.String()
+    use_permanent_title = fields.Bool()
+    max_displayed_items = fields.Integer(min=0)
+    items = fields.List(fields.Nested(DropdownItemSchema))
+    unusable = fields.Bool()
```
Comment (ajm13) [inline nanome/api/schemas/ui_schemas.py:363]:
```suggestion
```
dropdowns don't currently have an unusable state

File: nanome/api/schemas/ui_schemas.py [hunk 6]
```diff
@@ -0,0 +1,467 @@
...
+
+
+class MenuSchema(Schema):
+    is_menu = fields.Bool(default=True)
+    title = fields.String()
+    width = FloatRoundedField(max=1.0)
+    height = FloatRoundedField(max=1.0)
+    version = fields.Int(default=0)
+    effective_root = fields.Nested(LayoutNodeSchema, attribute='root')
+
```
Comment (ajm13) [inline nanome/api/schemas/ui_schemas.py:457]: missing enabled, index, locked

File: setup.cfg [hunk 7]
```diff
@@ -4,7 +4,7 @@ commit = True
 tag = True
 
 [metadata]
-description-file = README.md
+description_file = README.md
```
Comment (ajm13) [inline setup.cfg:7]:
```suggestion
description-file = README.md
```
this was correct before, to add the description to PyPI

File: nanome/api/schemas/structure_schemas.py [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+
+    @post_load
+    def make_atom(self, data, **kwargs):
+        new_obj = structure.Atom()
+        for key in data:
+            try:
+                setattr(new_obj, key, data[key])
+            except AttributeError:
+                raise AttributeError('Could not set attribute {}'.format(key))
+        return new_obj
```
Comment (ajm13) [inline nanome/api/schemas/structure_schemas.py:51]: I think for brevity, this should be refactored into a util method. I see this same for/try block repeated in all the schemas. something like:
```py
def make_atom(self, data, **kwargs):
    new_obj = init_obj(structure.Atom, data)
    # ... other schema-specific things
    return new_obj
```

File: setup.cfg (hunk 7 above)
Comment (mjrosengrant) [reply setup.cfg:7]: When testing the schema installs with setup.py, I got a warning saying the dash is deprecated, and to use underscore.

Comment (mjrosengrant) [reply setup.cfg:7]:
```
(venv) mike@MSI:~/workspace/nanome-lib$ python setup.py install 
/home/mike/workspace/nanome-lib/venv/lib/python3.7/site-packages/setuptools/dist.py:760: UserWarning: Usage of dash-separated 'description-file' will not be supported in future versions. Please use the underscore name 'description_file' instead
```

**Review by ajm13 — APPROVED** [review]

**Pull request merged**